s3/pkg/datastore/gcp/gcp.go

```diff
@@ -56,29 +64,65 @@ func (ad *GcsAdapter) Put(ctx context.Context, stream io.Reader, object *pb.Obje
 		limitedDataReader = stream
 	}
 
-	bucket := ad.session.NewBucket()
-	GcpObject := bucket.NewObject(bucketName)
-	d, err := ioutil.ReadAll(limitedDataReader)
-	data := []byte(d)
-	base64Encoded, hexEncoded := utils.Md5Content(data)
-	body := ioutil.NopCloser(bytes.NewReader(data))
-	err = GcpObject.Create(objectId, base64Encoded, "", size, body, models.Private)
+	// As per https://cloud.google.com/storage/docs/interoperability,
+	// using AWS CLINET SDK with HMAC credentials to upload object in GCS
+	// https://github.com/GoogleCloudPlatform/golang-samples
+	md5Writer := md5.New()
+	dataReader := io.TeeReader(limitedDataReader, md5Writer)
+	sess := session.Must(session.NewSession(&aws.Config{
+		Region:   aws.String("auto"),
+		Endpoint: aws.String("https://storage.googleapis.com"),
+		Credentials: credentials.NewStaticCredentials(
+			ad.backend.Access, ad.backend.Security, ""),
+	}))
+
+	// If the Storage Class is defined from the API request Header, set it else define defaults
+	var storClass string
+	var err error
+
+	if storageClass != "" {
+		storClass = storageClass
+	} else {
+		if object.Tier == 0 {
+			// default i.e STANDARD
+			object.Tier = utils.Tier1
+		}
+		storClass, err = osdss3.GetNameFromTier(object.Tier, utils.OSTYPE_GCS)
+		if err != nil {
+			log.Error("translate tier[%d] to gcp storage class failed", object.Tier)
+			return result, ErrInternalError
+		}
+	}
+	uploader := s3manager.NewUploader(sess)
+	input := &s3manager.UploadInput{
+		Body:         dataReader,
+		Bucket:       aws.String(bucketName),
+		Key:          aws.String(objectId),
+		StorageClass: aws.String(storClass),
+	}
+
+	log.Infof("Started uploading objectId:%s into GCS", objectId)
+	ret, err := uploader.Upload(input)
 	if err != nil {
-		log.Infof("put object[GCS] failed, object:%s, err:%v", objectId, err)
+		log.Errorf("Uplaoding objectId:%s failed with err:%v", objectId, err)
```

Comment:
nit: logging is inconsistent with rest of logging as it stats with block letter

Reply:
Changed